src/algopy_testing/op.py

```diff
@@ -938,69 +1000,315 @@ def _get_asset_holding(
+            raise ValueError(f"Invalid app index for apps in active transaction: {a}") from e
+        app_data = context.get_application(int(app_id))
+
+        if app_data is None:
+            return None, False
+
+        value = getattr(app_data, param, None)
+        return value, True
+
+    def app_approval_program(
+        self, a: algopy.Application | algopy.UInt64 | int, /
+    ) -> tuple[algopy.Bytes, bool]:
+        return self._get_app_param_from_ctx(a, "approval_program")
+
+    def app_clear_state_program(
+        self, a: algopy.Application | algopy.UInt64 | int, /
+    ) -> tuple[algopy.Bytes, bool]:
+        return self._get_app_param_from_ctx(a, "clear_state_program")
+
+    def app_global_num_uint(
+        self, a: algopy.Application | algopy.UInt64 | int, /
+    ) -> tuple[algopy.UInt64, bool]:
+        return self._get_app_param_from_ctx(a, "global_num_uint")
+
+    def app_global_num_byte_slice(
+        self, a: algopy.Application | algopy.UInt64 | int, /
+    ) -> tuple[algopy.UInt64, bool]:
+        return self._get_app_param_from_ctx(a, "global_num_byte_slice")
+
+    def app_local_num_uint(
+        self, a: algopy.Application | algopy.UInt64 | int, /
+    ) -> tuple[algopy.UInt64, bool]:
+        return self._get_app_param_from_ctx(a, "local_num_uint")
+
+    def app_local_num_byte_slice(
+        self, a: algopy.Application | algopy.UInt64 | int, /
+    ) -> tuple[algopy.UInt64, bool]:
+        return self._get_app_param_from_ctx(a, "local_num_byte_slice")
+
+    def app_extra_program_pages(
+        self, a: algopy.Application | algopy.UInt64 | int, /
+    ) -> tuple[algopy.UInt64, bool]:
+        return self._get_app_param_from_ctx(a, "extra_program_pages")
+
+    def app_creator(
+        self, a: algopy.Application | algopy.UInt64 | int, /
+    ) -> tuple[algopy.Account, bool]:
+        return self._get_app_param_from_ctx(a, "creator")
+
+    def app_address(
+        self, a: algopy.Application | algopy.UInt64 | int, /
+    ) -> tuple[algopy.Account, bool]:
+        return self._get_app_param_from_ctx(a, "address")
 
 
 AppParamsGet = _AppParamsGet()
 
 
 class _AppLocal:
-    def __getattr__(self, name: str) -> Any:
-        raise NotImplementedError(
-            f"AppLocal.{name} is currently not available as a native "
-            "`algorand-python-testing` type. Use your own preferred testing "
-            "framework of choice to mock the behaviour."
+    def _get_local_state(self, key: bytes) -> algopy.LocalState[Any]:
+        from algopy_testing import get_test_context
+
+        test_context = get_test_context()
+        if not test_context or not test_context._active_contract:
+            raise ValueError("No active contract or test context found.")
+
+        local_states = test_context._active_contract._get_local_states()
+        local_state = local_states.get(key)
+        if local_state is None:
+            raise ValueError(f"Local state with key {key.decode()} not found")
```

Comment:
this could cause another error if key is not a valid UTF8 sequence of bytes

Reply:
Resolved, feel free to reopen if you think something else is missing